Add AppUser class(Extened from IdentityUser)

ewgSht opened the pull request:
Add Controller for work with AppUser(CRUD)

ewgSht committed:
Add AppUser class(Extened from IdentityUser)
Add Controller for work with AppUser(CRUD)

Ibraga96 commented on UtopiaCity/Controllers/CitizenAccount/CitizensAccountController.cs:
```diff
@@ -0,0 +1,91 @@
+        }
+
+        [HttpGet]
+        public async Task<IActionResult> EditAsync(string id)
```
can't understand for what you're using id in this method, and for what you' ve created service class"CitizenAccount" if you're not using it , and how do you understand which model of User will be edited ?

ewgSht replied:
Hi!
1. in this method, I use a parameter to check for the occupancy of this parameter, just as we did in the lesson when working with the emergencyreport controller. That is, if we pass the parameter in the form https://localhost:44375/CitizensAccount/Edit/83b43606-a779-474f-8e83-9346825739d1 everything will be ok, the method will work. And if we pass https://localhost:44375/CitizensAccount/Edit the method does not understand which representative of the User class to work with. But in this case, the entire controller has the Authorize attribute, and in principle, we will not get into this method without authorization. So perhaps this parameter is unnecessary.

2. I created this class for other developers, I saw use-case where other developer need find User by Id for their use-case. And also, I think if I will must append add functionality it will be easier.I don't use functionality of class "CitizensAccountService" because class "AppUser" extended from "IdentityUser" and for this class microsoft created perfect methods and I use them.

3. How I wrote before If U want use methods of class "CitizensAccountController" U must be authorized. And all controllers has "User" property. And this User is exactly the one that is needed.

ewgSht replied:
Ibraga96, 
While I was answering you,I think about parameter ID in Edit method, And I think it redundant.
Thanks:)

Ibraga96 replied:
no problems, if i understood correctly you have extended, templated class User , and use this class, and also you'll have your own authorization form . correct me if i wrong

ewgSht replied:
Yes everything right. But only I don't have specific authorization form, it is standart form by microsoft

mainstreetswagger replied:
> Hi!
> 
> 1. in this method, I use a parameter to check for the occupancy of this parameter, just as we did in the lesson when working with the emergencyreport controller. That is, if we pass the parameter in the form https://localhost:44375/CitizensAccount/Edit/83b43606-a779-474f-8e83-9346825739d1 everything will be ok, the method will work. And if we pass https://localhost:44375/CitizensAccount/Edit the method does not understand which representative of the User class to work with. But in this case, the entire controller has the Authorize attribute, and in principle, we will not get into this method without authorization. So perhaps this parameter is unnecessary.
> 2. I created this class for other developers, I saw use-case where other developer need find User by Id for their use-case. And also, I think if I will must append add functionality it will be easier.I don't use functionality of class "CitizensAccountService" because class "AppUser" extended from "IdentityUser" and for this class microsoft created perfect methods and I use them.
> 3. How I wrote before If U want use methods of class "CitizensAccountController" U must be authorized. And all controllers has "User" property. And this User is exactly the one that is needed.

we can find a current user.
var currentUser = User.Identity.Name;
And name can store that user id.

mainstreetswagger replied:
we can find a current user.
var currentUser = User.Identity.Name;
And name can store that user id.

mainstreetswagger replied:
we can assign id to Claims. and claims pass to token. Tokens configured in StartUp.

ewgSht committed:
Add CitizensTask controller for work with citizen tasks
add base crud operations
add views

ewgSht committed:
Add "CitizenTaskService" class for working with basic crud operation for controller
Update "CitizenTaskController" for work with "CitizenTaskService"
Add "CitizenTaskInitializer" for initialize CitizenTask table in database
Update some views

mainstreetswagger commented on UtopiaCity/Data/Migrations/20210802173202_CitizenAcc1.cs:
```diff
@@ -0,0 +1,73 @@
+        protected override void Up(MigrationBuilder migrationBuilder)
+        {
+            migrationBuilder.CreateTable(
+                name: "CitizenAccUsers",
```
Is it AppUser? If yes how was the name changed to CitizensAccuser?

mainstreetswagger replied:
Ok. Whatever you drop it anyway in next migration I see.

mainstreetswagger commented on UtopiaCity/Data/Migrations/20210804064209_Update_IdentityUser_V1.cs:
```diff
@@ -0,0 +1,80 @@
+                name: "Surname",
+                table: "AspNetUsers",
+                nullable: true);
+        }
```
Finally, you added that fields to IdentityUser, so no need for AppUser model because you can access them in IdentityUser.

mainstreetswagger commented on UtopiaCity/Data/Migrations/20210804064209_Update_IdentityUser_V1.cs:
```diff
@@ -0,0 +1,80 @@
+                name: "Balance",
+                table: "AspNetUsers",
+                nullable: true);
+
```
how did you manage to add new fields to AspNetUsers?

mainstreetswagger replied:
As I know efcore needs  that altered model AspNetUser and where is that model.

mainstreetswagger reviewed (commented):
in the end it works fine.

crocomoth commented on UtopiaCity/Controllers/CitizenAccount/CitizensAccountController.cs:
```diff
@@ -0,0 +1,87 @@
+    public class CitizensAccountController : Controller
+    {
+        private readonly ApplicationDbContext _dbContext;
+        private UserManager<AppUser> _userManager;
```
can also be readonly?

crocomoth commented on UtopiaCity/Controllers/CitizenAccount/CitizensAccountController.cs:
```diff
@@ -0,0 +1,87 @@
+        private readonly ApplicationDbContext _dbContext;
+        private UserManager<AppUser> _userManager;
+        private readonly SignInManager<AppUser> _signInManager;
+        public CitizensAccountController(ApplicationDbContext dbContext, UserManager<AppUser> userManager, SignInManager<AppUser> signInManager)
```
empty line

crocomoth commented on UtopiaCity/Models/CitizenAccount/Gender.cs:
```diff
@@ -0,0 +1,13 @@
+{
+    public enum Gender
+    {
+        Male=1,
```
0 and 1 could be better, since logic will break if for some reason we get 0 from front, even if it help to detect stuff

crocomoth commented on UtopiaCity/Views/EmergencyReport/EditEmergencyReportView.cshtml:
```diff
@@ -34,6 +34,6 @@
 </div>
 
-@section Scripts {
+@*@section Scripts {
```
?

ewgSht committed:
Update CitizensAccountController.cs

Updated the controller class(set the readonly field).

ewgSht committed:
Update CitizensAccountController.cs

ewgSht committed:
Update Gender.cs

ewgSht committed:
Update EditEmergencyReportView.cshtml

ewgSht committed:
Merge branch 'dev' into CitizenAccount-2

ewgSht committed:
Merge branch 'dev' into CitizenAccount-2

mainstreetswagger reviewed (approved):
Ok.  Make sure all conflicts resolved before merge.

ewgSht committed:
Merge branch 'dev' into CitizenAccount-2
